internal/contract/infrastructure_machine.go

```diff
@@ -0,0 +1,79 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package contract
+
+import (
+	"sync"
+)
+
+// InfrastructureMachineContract encodes information about the Machine API contract for InfrastructureMachine objects
+// like DockerMachines, AWS Machines, etc.
+type InfrastructureMachineContract struct{}
+
+var infrastructureMachine *InfrastructureMachineContract
+var onceInfrastructureMachine sync.Once
+
+// InfrastructureMachine provide access to the information about the Machine API contract for InfrastructureMachine objects.
+func InfrastructureMachine() *InfrastructureMachineContract {
+	onceInfrastructureMachine.Do(func() {
+		infrastructureMachine = &InfrastructureMachineContract{}
+	})
+	return infrastructureMachine
+}
+
+// Ready provides access to status.ready field in an InfrastructureMachine object.
+func (m *InfrastructureMachineContract) Ready() *Bool {
+	return &Bool{
+		path: []string{"status", "ready"},
+	}
+}
+
+// FailureReason provides access to the status.failureReason field in an InfrastructureMachine object. Note that this field is optional.
+func (m *InfrastructureMachineContract) FailureReason() *String {
+	return &String{
+		path: []string{"status", "failureReason"},
+	}
+}
+
+// FailureMessage provides access to the status.failureMessage field in an InfrastructureMachine object. Note that this field is optional.
+func (m *InfrastructureMachineContract) FailureMessage() *String {
+	return &String{
+		path: []string{"status", "failureMessage"},
+	}
+}
+
+// Addresses provides access to the status.addresses field in an InfrastructureMachine object. Note that this field is optional.
+func (m *InfrastructureMachineContract) Addresses() *MachineAddresses {
```

Comment:
The [contract](https://cluster-api.sigs.k8s.io/developer/providers/machine-infrastructure.html) specifies that `status.addresses` field is a list of MachineAddress structs. Is there a way to access indices in a list using the JSON path syntax we've outlined here? Currently I've just made status.addresses a leaf node and that returns a new MachineAddress type I've defined. I'm definitely open to suggestions on how to handle lists in the contract. While there is an unstructured.NestedSlice method, I had to define a new type to implement the Set() and Get() methods.

Reply:
Same as in https://github.com/kubernetes-sigs/cluster-api/pull/7138#discussion_r976269146

I would just let Addresses() get/set the entire slice. It's a lot easier to handle then reimplemting slice insert/remove/update in the contract package